### PR TITLE
[IIO] test cases for state change and corresponding bug fixes @open sesame 03/27 18:44

### PR DESCRIPTION
**Pull request by kparichay**
1. Created virtual iio device to test against
2. Added test case to test state change from NULL to PLAYING and backwards
3. Bug fixes-
  3.1 setting of fixated caps moved to _fixating() from _start()
  3.2 channel property list memory issue resolved
  3.3 missing sampling frequency file while setting it is now a warning
      skip to default frequency if file is not present
  3.4 More minor bug fixes
4. IIO base directories moved to header file
  These directories are modified to temporary location when testing with simulated device

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 1]
```diff
@@ -1636,9 +1650,22 @@ gst_tensor_src_iio_fixate (GstBaseSrc * src, GstCaps * caps)
...
+  if (gst_caps_can_intersect (caps, self->fixated_caps)) {
+    updated_caps = gst_caps_intersect (caps, self->fixated_caps);
+  } else {
+    GST_ERROR_OBJECT (self,
+        "No intersection between generic and fixated caps of the element.");
+  }
+  gst_caps_unref (caps);
+  gst_caps_unref (self->fixated_caps);
+
+  self->fixated_caps = gst_caps_ref (updated_caps);
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: You don't need to ref the returned caps of gst_caps_can_intersect.

Refer: Line 1499 - 1583 of https://github.com/alessandrod/gstreamer/blob/master/gst/gstcaps.c

Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Yes, gst_caps_intersect() increases the refcount before returning.

updated_caps is passed to gst_caps_fixate() (Line 1668), which takes the ownership of the passed caps.
`self->fixated_caps = gst_caps_ref (updated_caps);` intends to store the latest caps for later use (if needed)
That's why I use gst_caps_ref() - if pad unref the caps, self->fixated_caps should still remain valid.

Thanks to your comment, i found 2 bugs 
- caps returned from `gst_caps_fixate (updated_caps)` should be stored, as it is latest and finalized caps
- self->fixated_caps should call 'gst_caps_unref()' in _stop()

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 2]
```diff
@@ -1346,7 +1364,7 @@ gst_tensor_src_iio_start (GstBaseSrc * src)
 
   /** get all the channels that exist and then set enable on them */
   dirname = g_build_filename (self->device.base_dir, CHANNELS, NULL);
-  guint num_channels_enabled =
+  gint num_channels_enabled =
```
Comment (wooksong) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: After merging https://github.com/nnsuite/nnstreamer/pull/1213, this declaration may not be allowed.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 3]
```diff
@@ -354,6 +348,7 @@ gst_tensor_src_iio_channel_properties_free (gpointer data)
   g_free (prop->generic_name);
   g_free (prop->base_dir);
   g_free (prop->base_file);
+  g_free (prop);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: plz check the function gst_tensor_channel_list_filter_enabled()
```
    gst_tensor_src_iio_channel_properties_free (channel);
    *list_addr = g_list_remove (list, channel);
```

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 4]
```diff
@@ -1313,23 +1320,34 @@ gst_tensor_src_iio_start (GstBaseSrc * src)
       gst_tensor_src_iio_set_frequency (self->device.base_dir,
       self->sampling_frequency);
   if (0 == sampling_frequency) {
-    GST_ERROR_OBJECT (self, "IIO device does not support %lu frequency.\n",
-        (gulong) self->sampling_frequency);
+    GST_ERROR_OBJECT (self, "Error in setting frequency for device %s.\n",
+        self->device.name);
     goto error_trigger_free;
+  } else if (0 == sampling_frequency) {
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: invalid condition (line 1322)

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 2 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: done.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 3 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Good catch, done.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 4 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: updated.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 1 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Moved fixating of caps to _fixate(). This removes unnecessary storage of fixated_caps and calculating fixated caps in _fixate() makes code more readable
This also automatically resolves the two bugs mentioned above.

File: gst/nnstreamer/tensor_source/tensor_src_iio.h [hunk 5]
```diff
@@ -31,6 +31,7 @@
 #include <gst/base/gstbasesrc.h>
 #include <tensor_common.h>
 #include <poll.h>
+#include <stdio.h>
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.h]: Is this required?

Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.h]: Currently it is being used for polling on a file descriptor and non-blocking reading of file.
However, this is not necessary and possibly be worked around with glib functionality.
Will remove it in another PR (issue ) as this PR just moves `#include <stdio.h>` from tensor_src_iio.c to tensor_src_iio.h

**Pull request merged**